### DISCUSSION TITLE
Per message context for async socketio client app

### DISCUSSION THREAD

**Discussion by bhakta0007** [miguelgrinberg/python-socketio#741] [Q&A]
This is a question:

I am using asyncio + socketio[client] in python 3.6. The program connects to a socketio server (hub). It is possible for my program to get multiple concurrent requests, each of which carries a context (say inside the payload of the incoming message).

```
@sio.on('message')
async def messageHandler(data):
    ctx = data.get("context", {})
    # set the context so any functions under this calling tree can access the context.
    await F1()

async def F1():
   logger.info("Entering F1 with context: {}".format(HOW_DO_I_GET_THE_CTX?))
   await F2()

async def F2():
   logger.info("Entering F2 with context: {}".format(HOW_DO_I_GET_THE_CTX?))
```

eventually, I want this context to appear in JSON log (I will set up a JSON formatted that picks a dynamic context field and logs it)., so I can push it to ElasticSearch and be able to search based on the context-id (think transaction tracing).

I explored contextvars, but I see the context getting overwritten with the latest request. 

Like how flask has request context, is there a way to associate context to every incoming async socketio message ? This context can be something we pull from payload or something we create on the fly (say unique request-Id). The context must be derivable from any functions that are called in the tree, like the log formatter.

**Comment by miguelgrinberg** [comment]
I'm not sure I really understand what you want, but Flask application and request contexts are based on the [contextvar](https://docs.python.org/3/library/contextvars.html) objects from Python. So maybe that is what you want?

**Comment by bhakta0007** [comment]
I had tried context vars, but what I hadn't realized was that I was on python3.6 and I had done a "pip install contextvars", with that package imported, things didn't work as expected.

I moved to 3.7 and contextvars work as expected.

**Comment by mozartilize** [comment]
Hi @miguelgrinberg I think I don't fully understand contextvar but can I use it for each event of same sid like this?

```python
some_ctx = contextvars.ContextVar('some_ctx')

@sio.event
async def connect(sid, eviron, auth):
    result = get_some_ctx(auth)
    some_ctx.set(result)


@sio.event
async def disconnect(sid):
    result = some_ctx.get()
    # some code goes here
```

I know there's `sio.session` but how does it scale if I run on gunicorn with multiple uvicorn workers?

**Comment by miguelgrinberg** [comment]
@mozartilize Each event handler runs in the context of a new async task, so no, it is not the appropriate use for what you want to do. The user sessions is what you want.

**Comment by mozartilize** [comment]
@miguelgrinberg thank you, how about its scale if I run on gunicorn with multiple uvicorn workers?

**Comment by miguelgrinberg** [comment]
I don't understand what do you mean by scale. Using a contextvar you seemed okay storing one value per client, so how is it different with a session?

**Comment by mozartilize** [comment]
yeah, I'm messed up. Look like there's nothing I can do but putting more ram in it if thousands of users connect to my service :D

**Comment by miguelgrinberg** [comment]
@mozartilize Well, if you need to store a lot of data for each user, then you can use a database. Or Redis. Then the only thing you need to save in memory is the key under which the data for each user is stored in your db.
